### ISSUE TITLE
Writer error "unexpected EOF" with v0.4.2 + SASL + TLS

### ISSUE THREAD

**Issue by pierrre** [segmentio/kafka-go#499]
**Describe the bug**
When I try to write a message, I receive "unexpected EOF" error message

**Kafka Version**
I'm using the version available from confluent.cloud, so I think it's the latest.

**To Reproduce**
```go
package main

import (
	"context"
	"crypto/tls"
	"log"
	"time"

	"github.com/segmentio/kafka-go"
	"github.com/segmentio/kafka-go/sasl/plain"
)

func main() {
	ctx := context.Background()
	w := kafka.NewWriter(kafka.WriterConfig{
		Brokers: []string{
			"xxx.gcp.confluent.cloud:9092",
		},
		Topic: "test",
		Dialer: &kafka.Dialer{
			Timeout:   10 * time.Second,
			DualStack: true,
			TLS:       &tls.Config{},
			SASLMechanism: plain.Mechanism{
				Username: "yyy",
				Password: "zzz",
			},
		},
		BatchSize: 1,
		Logger: kafka.LoggerFunc(func(format string, args ...interface{}) {
			log.Printf("Kafka writer: "+format, args...)
		}),
	})
	for {
		err := w.WriteMessages(ctx, kafka.Message{
			Value: []byte(time.Now().Format(time.RFC3339)),
		})
		if err != nil {
			panic(err)
		}
		log.Println("message sent")
		time.Sleep(1 * time.Second)
	}
}
```

The call to `w.WriteMessages` returns an error.

**Expected behavior**
I shouldn't not receive any error

**Additional context**
I'm using v0.4.2 with TLS + SASL auth.

It was working fine with 0.3.8 and 0.4.0.
0.4.1 has a panic.

I think it's actually the same issue as https://github.com/segmentio/kafka-go/issues/490
But I don't understand why it was closed, the issue is not fixed.
Someone said https://github.com/segmentio/kafka-go/issues/490#issuecomment-676238806
but it's not true, right ?

**Comment by pierrre** [comment]
PS: while testing on a different Kafka cluster, I was able to access the error log on the broker side:
```
[2020-08-31 14:02:52,081] INFO [SocketServer brokerId=1] Failed authentication with /1.2.3.4 (Invalid SASL/PLAIN response: expected 3 tokens, got 4) (org.apache.kafka.common.network.Selector)
```

I'm not sure, but it could be related to this code:
https://github.com/segmentio/kafka-go/blob/3d2413fa36e84d1f459dc86f88e5ad7652996c52/sasl/plain/plain.go#L23
However I don't know from where the 3rd `\x00` is written :shrug:

**Comment by tarnowsc** [comment]
The same issue occurs when using:
```
scram.Mechanism(scram.SHA256, user, password)
```

**Comment by swarupdonepudi** [comment]
I get the same exact error. Watching for updates. Also, I see that NewWriter is now deprecated? can someone please share an example of instantiating a writer in v0.4.2?

**Comment by pierrre** [comment]
@swarupdonepudi thank you for noticing this !
But the issue is still the same.
Here is my new code:
```go
package main

import (
	"context"
	"crypto/tls"
	"log"
	"time"

	"github.com/segmentio/kafka-go"
	"github.com/segmentio/kafka-go/sasl/plain"
)

func main() {
	ctx := context.Background()
	w := &kafka.Writer{
		Addr:  kafka.TCP("xxx.gcp.confluent.cloud:9092"),
		Topic: "test",
		Transport: &kafka.Transport{
			TLS: &tls.Config{},
			SASL: plain.Mechanism{
				Username: "yyy",
				Password: "zzz",
			},
		},
		BatchSize: 1,
		Logger: kafka.LoggerFunc(func(format string, args ...interface{}) {
			log.Printf("Kafka writer: "+format, args...)
		}),
	}
	for {
		err := w.WriteMessages(ctx, kafka.Message{
			Value: []byte(time.Now().Format(time.RFC3339)),
		})
		if err != nil {
			panic(err)
		}
		log.Println("message sent")
		time.Sleep(1 * time.Second)
	}
}
```

**Comment by vtolstov** [comment]
i think that this bug in 0.4.2 release that is fixed in master. I'm also have test service that uses 0.3.8 with sasl auth and it works fine, but in 0.4.2 it fails to publish message. Try to use master branch of kafka-go.

i think that we need to have next release like 0.4.3 to have this issue fixed

**Comment by pierrre** [comment]
@vtolstov correct me if I'm wrong, but I think the master branch is equal to the tag v0.3.9 currently.
So it doesn't have all the new changes from the `0.4` branch.

**Comment by vtolstov** [comment]
ah, if you need 0.4.x branch, yes, you need to wait new 0.4 release

**Comment by zzzhr1990** [comment]
Now the release is v0.4.5, this bug still not fixed.

**Comment by zzzhr1990** [comment]
@scottwhite However, 0.3.x works fine

**Comment by icrustandi** [comment]
I can confirm we're having the same issues here with a confluent broker

**Comment by zzzhr1990** [comment]
Any fix threre?

**Comment by jedna** [comment]
Works again for me with 0.4.6.

**Comment by pierrre** [comment]
probably fixed by #541

**Comment by pierrre** [comment]
I can confirm, it's fixed in v0.4.6

**Comment by achille-roussel** [comment]
Thanks for verifying @pierrre, and thanks everyone for your reports!

**Comment by michael-joseph-payne** [comment]
I ran into this behavior again while trying to hook up a new broker in Confluent Cloud.  All attempts at sending a message with a Writer resulted in an unexpected EOF error.  Eventually, I isolated it down to a struct field in Transport.  We are using v0.4.27 kafka-go.

Per the example documentation, we started out using (and had the errors mentioned above):

sharedTransport := &kafka.Transport{
		SASL: mechanism,
	}

I went through the Confluent documentation and also took a look at the default config they supply.  The clue that tipped me off was that they require TLS 1.2 minimum.  Adding that to our Transport struct as below solved the issue for us:

sharedTransport := &kafka.Transport{
		SASL: mechanism,
		// If we don't include this we get expected EOF errors
		TLS: &tls.Config{
			MinVersion: tls.VersionTLS12,
		},
	}
